// app/helpers/EmailHelper.js
/**
 * Created by shadygabal on 12/24/15.
 */

var emailHelper = (function() {
    var nodemailer = require('nodemailer');
    var xoauth2 = require('xoauth2');
    var ErrorCodes = require('app/helpers/ErrorCodes.js');

    var generator = xoauth2.createXOAuth2Generator({
        user: 'info@experiencesoiree.com',
        //pass: process.env.EMAIL_PW,
        clientId: process.env.EMAIL_CLIENT_ID,
        clientSecret: process.env.EMAIL_CLIENT_SECRET,
        refreshToken: process.env.EMAIL_REFRESH_TOKEN
        //accessToken: process.env.EMAIL_ACCESS_TOKEN
    });

    var transporter;

    generator.getToken(function(user, token, accessToken){
        transporter = nodemailer.createTransport({
            service: 'gmail',
            auth: {
                xoauth2: generator
            }
        });
    });

    // create reusable transporter object using SMTP transport
    //var

    var validator = require('validator');

    return {
        sendVerificationEmail: function (email, user, successCallback, errorCallback) {
            if (!user){
                return errorCallback(ErrorCodes.MissingData);
            }

            //user.generateVerificationCode();
            var code = user.verificationCode;
            if (!code){
                user.generateVerificationCode();
            }

            // setup e-mail data with unicode symbols
            var mailOptions = {
                //from: 'Soirée <donotreply@experiencesoiree.com>', // sender address
                to: email, // list of receivers
                subject: "Start experiencing Soirée", // Subject line
                text: "Hey " + user.firstName + ",\n\n" +
                      "It's time to begin experiencing Soirée. Enter verification code \n"  +
                      code +"\n to get started.", // plaintext body
                html: "Hey " + user.firstName + ",<br /><br />" +
                      "It's time to begin experiencing Soirée. Enter verification code <br /><br />"  +
                      "<b> " + code + " </b> <br /><br /> to get started."// html body
            };


            // send mail with defined transport object
            transporter.sendMail(mailOptions, function(error, info){
                if(error){
                    console.log(error);
                    errorCallback(error);
                }
                else {
                    console.log('Message sent: ' + info.response);
                    user.pendingVerification = true;
                    user.save();
                    successCallback();
                }

            });
        },

        sendBetaSignupEmail : function(email, gender, os){
            var body, html;
            if (os === 'ios'){
                html = "Thanks! You're signed up for the Soirée beta. We're currently working really hard (with far too much caffeine in our system) to make sure that the app and the experience are perfect. We expect the beta to start on <strong> June 24th </strong>, and we'll email you soon with a link to download the app. Can't wait to see you there!<br /><br /><strong>xoxo,<br />the people at Soirée</strong>";
                body = "Thanks! You're signed up for the Soirée beta. We're currently working really hard (with far too much caffeine in our system) to make sure that the app and the experience are perfect. We expect the beta to start on June 24th, and we'll email you soon with a link to download the app. Can't wait to see you there!\n\nxoxo,\nthe people at Soirée";
            }
            else{
                html = "Thanks! You're signed up for the Soirée beta. Unfortunately, the beta that is coming up is currently only for iOS. However, our coders are currently working 25 hours a day 8 days a week to make sure that we launch our android version really soon. We'll send you an email when it's time for the android beta. Can't wait to see you there!<br /><br /><strong>xoxo,<br />the people at Soirée</strong>";
                body = "Thanks! You're signed up for the Soirée beta. Unfortunately, the beta that is coming up is currently only for iOS. However, our coders are currently working 25 hours a day 8 days a week to make sure that we launch our android version really soon. We'll send you an email when it's time for the android beta. Can't wait to see you there!\n\nxoxo,\nthe people at Soirée";
            }


            var mailOptions = {
                from: 'Soirée <info@experiencesoiree.com>', // sender address
                to: email, // list of receivers
                subject: "You're signed up for the Soirée beta!", // Subject line
                text: body,
                html : html
            };


            // send mail with defined transport object
            transporter.sendMail(mailOptions, function(error, info){
                if(error){
                    console.log(error);
                    //errorCallback(error);
                }
                else {
                    console.log('Email sent to beta signupee: ' + email);
                    //successCallback();
                }

            });
        },

        sendBusinessRegisteredEmail : function(business, password, successCallback, errorCallback){
            var html = "Thanks! Your business, " + business.businessName + ", is now signed up with Soirée! Here is your login information: <br/><br/>" +
                    "username : <strong>" + business.email + "</strong><br /> " +
                    "password : <strong> " + password + "</strong>" +
                "<br /><br/> Please save this password for your records. <br/> We recommend deleting this email and keeping this password somewhere only you have access to.";

            var body = "Thanks! Your business, " + business.businessName + ", is now signed up with Soirée! Here is your login information: \n\n" +
                    "username : " + business.email + "\n"
                    "password : " + password + "\n" +
                    "\n\n Please save this password for your records. \n We recommend deleting this email and keeping this password somewhere only you have access to.";

            var subject = "Your Soirée login information";
            this.sendMail(business.email, subject, body, html, successCallback, errorCallback);

        },

        sendMail : function(email, subject, body, html, successCallback, errorCallback){
            if (!email || !body || !html) return errorCallback(ErrorCodes.MissingData);

            var mailOptions = {
                from: 'Soirée <businesses@experiencesoiree.com>', // sender address
                to: email, // list of receivers
                subject: subject, // Subject line
                text: body,
                html : html
            };

            // send mail with defined transport object
            transporter.sendMail(mailOptions, function(error, info){
                if(error){
                    console.log(error);
                    errorCallback(ErrorCodes.Error);
                }
                else {
                    console.log('Email sent to : ' + email);
                    successCallback();
                }
            });
        },

        validateEmail: function(email){
            return validator.isEmail(email);
        }
    }

}());

module.exports = emailHelper;
