// app/helpers/EmailHelper.js
/**
 * Created by shadygabal on 12/24/15.
 */

var emailHelper = (function() {
    var nodemailer = require('nodemailer');
    var xoauth2 = require('xoauth2');
    var ErrorCodes = require('app/helpers/ErrorCodes.js');
    var Globals = require('app/helpers/Globals');

    var generator = xoauth2.createXOAuth2Generator({
        user: 'info@experiencesoiree.com',
        clientId: process.env.EMAIL_CLIENT_ID,
        clientSecret: process.env.EMAIL_CLIENT_SECRET,
        refreshToken: process.env.EMAIL_REFRESH_TOKEN
    });

    var transporter;

    generator.getToken(function(user, token, accessToken){
        transporter = nodemailer.createTransport({
            service: 'gmail',
            auth: {
                xoauth2: generator
            }
        });
    });

    // create reusable transporter object using SMTP transport
    //var

    var validator = require('validator');

    return {
        sendVerificationEmail: function (email, user, successCallback, errorCallback) {
            if (!user){
                return errorCallback(ErrorCodes.MissingData);
            }

            var code = user.verificationCode;
            if (!code){
                user.generateVerificationCode();
            }



            var body = "Hey " + user.firstName + ",\n\n" +
                "It's time to begin experiencing Soirée. Enter verification code \n"  +
                code +"\n to get started.";
            var html = "Hey " + user.firstName + ",<br /><br />" +
                "It's time to begin experiencing Soirée. Enter verification code <br /><br />"  +
                "<b> " + code + " </b> <br /><br /> to get started.";


            this.sendMail(email, "You're signed up for the Soirée beta!", body, html, successCallback, errorCallback);


        },

        sendBetaSignupEmail : function(email, gender, os){
            var body, html;
            if (os === 'ios'){
                html = "Thanks for signing up for Soiree - the future of making real-world friends and memories. We are putting the final touches on our beta and will be launching any day now, so stay tuned!<br /><br /><strong>xoxo,<br />your friends at Soirée</strong><br /><br />" +
                    "Interested in partnering with Soiree? <br /> Join the many businesses that are lining up to be a part of the future of safe, real connections. Email us at: <a href=\"mailto:business@experiencesoiree.com\">business@experiencesoiree.com</a>";
                body = "Thanks for signing up for Soiree - the future of making real-world friends and memories. We are putting the final touches on our beta and will be launching any day now, so stay tuned!\n\nxoxo,\nyour friends at Soirée\n\n" + "Interested in partnering with Soiree? \n Join the many businesses that are lining up to be a part of the future of safe, real connections. Email us at: business@experiencesoiree.com";
            }
            else{
                html = "Thanks for signing up for Soiree - the future of making real-world friends and memories. Unfortunately, our upcoming beta is only for iOS (we <3 our Android people, but we couldn't create two apps simultaneously and have them both have the awesome experience we want for our users). <br /><br /> However, as we speak, an overworked coder somewhere in his parents' basement is working on the Android app, so stay tuned!<br /><br /><strong>xoxo,<br />your friends at Soirée</strong><br /><br />" +
                    "Interested in partnering with Soiree? <br /> Join the many businesses that are lining up to be a part of the future of safe, real connections. Email us at: <a href=\"mailto:business@experiencesoiree.com\">business@experiencesoiree.com</a>";
                body = "Thanks for signing up for Soiree - the future of making real-world friends and memories. Unfortunately, our upcoming beta is only for iOS (we <3 our Android people, but we couldn't make two apps simultaneously and have them both have the awesome experience we want). \n\n However, as we speak, an overworked coder somewhere in his parents' basement is working on the Android app, so stay tuned!\n\nxoxo,\n your friends at Soirée \n\n" +
                    "Interested in partnering with Soiree? <br /> Join the many businesses that are lining up to be a part of the future of safe, real connections. Email us at: <a href=\"mailto:business@experiencesoiree.com\">business@experiencesoiree.com</a>";
            }

            this.sendMail(email, "You're signed up for the Soirée beta!", body, html);

        },

        sendBusinessRegisteredEmail : function(business, password, successCallback, errorCallback){
            var html = "Thanks! Your business, " + business.businessName + ", is now signed up with Soirée! Here is your login information: <br/><br/>" +
                    "username : <strong>" + business.email + "</strong><br /> " +
                    "password : <strong> " + password + "</strong>" +
                "<br /><br/> Please save this password for your records. <br/> We recommend deleting this email and keeping this password somewhere only you have access to.";

            var body = "Thanks! Your business, " + business.businessName + ", is now signed up with Soirée! Here is your login information: \n\n" +
                    "username : " + business.email + "\n"
                    "password : " + password + "\n" +
                    "\n\n Please save this password for your records. \n We recommend deleting this email and keeping this password somewhere only you have access to.";

            var subject = "Your Soirée login information";
            this.sendMail(business.email, subject, body, html, successCallback, errorCallback);

        },

        sendMail : function(email, subject, body, html, successCallback, errorCallback){
            if (!email || !body || !html) return errorCallback(ErrorCodes.MissingData);
            if (Globals.test){
                if (successCallback) return successCallback();
                return;
            }

            var mailOptions = {
                from: 'Soirée <businesses@experiencesoiree.com>', // sender address
                to: email, // list of receivers
                subject: subject, // Subject line
                text: body,
                html : html
            };

            // send mail with defined transport object
            transporter.sendMail(mailOptions, function(error, info){
                if(error){
                    console.log(error);
                    errorCallback(ErrorCodes.Error);
                }
                else {
                    console.log('Email sent to : ' + email);
                    successCallback();
                }
            });
        },

        validateEmail: function(email){
            return validator.isEmail(email);
        }
    }

}());

module.exports = emailHelper;
