// app/helpers/Globals.js
var env = require('get-env')({
    development : ['dev', 'development'],
    production : ['prod', 'production']
});

var ResHelper = require('./ResHelper.js');


var Globals = {
    io : {},
    socketsInRoom : function(roomId) {
        var arr = [], room = this.io.sockets.adapter.rooms[roomId];
        if (room) {
            for (var id in room) {
                arr.push(io.sockets.adapter.nsp.connected[id]);
            }
        }
        return arr;
    },
    development: env === 'development',
    production : env === 'production',
    saveErrorCallback : function(err){
        if (err){
            console.log(err);
        }
    },
    fetchErrorCallback : function(res){
        return function(err){
            ResHelper.sendError(res, error);
        };
    },
    SOIREE : "Soirée",
    SOIREE_LOWERCASE : "soirée",
    RETRIES_ON_DUPLICATE_CONFIRMATION_CODE: 5,
    RETRIES_ON_DUPLICATE_ID : 10,
    genders : ["male", "female"],
    providers : ["facebook", "userpw"],
    soireeTypes : ["Dinner", "Drinks", "Speed Date", "TEST"],
    initialChargePerSoireeType : {"Dinner" : 99, "Drinks" : 99, "Speed Date" : 99, "TEST" : 99},
    availableTimesPerSoireeType : {
        "Dinner" : ["6:00 PM", "6:30 PM", "7:00 PM", "7:30 PM"],
        "Drinks" : ["9:00 PM", "9:30 PM", "10:00 PM", "10:30 PM", "11:59 PM", "3:00 AM", "5:00 AM"],
        "Speed Date" : ["7:00 PM", "8:00 PM", "9:00 PM", "1:00 AM", "5:00 AM"],
        "TEST" : ["1:00 PM", "2:00 PM", "11:00 PM", "12:00 AM", "1:00 AM", "2:00 AM", "3:00 AM", "4:00 AM", "5:00 AM", "6:00 AM", "7:00 AM"]
    },
    numUsersMaxPerSoireeType : {"Dinner" : 10, "Drinks" : 20, "Speed Date" : 40, "TEST" : 3},
    numUsersMinPerSoireeType : {"Dinner" : 5, "Drinks" : 8, "Speed Date" : 15, "TEST" : 2},
    scheduledSoireeAvailableTimes : [
        "11:00 AM","11:30 AM","12:00 PM","12:30 PM","1:00 PM",
        "1:30 PM","2:00 PM","2:30 PM","3:00 PM","3:30 PM",
        "4:00 PM","4:30 PM", "5:00 PM", "5:30 PM", "6:00 PM",
        "6:30 PM", "7:00 PM", "7:30 PM", "8:00 PM","8:30 PM",
        "9:00 PM","9:30 PM","10:00 PM","10:30 PM","11:00 PM"
    ],
    soireesCancelling : {},
    minMaleToFemaleRatios : {//max num of one gender per another
        "Dinner": "4/2",
        "Drinks": "4/2",
        "Speed Date": "1/1",
        "TEST" : "4/2"
    },
    numSoireesHostedPerBusiness : {}

};


module.exports = Globals;
