// app/helpers/Globals.js
var env = require('get-env')({
    development : ['dev', 'development'],
    production : ['prod', 'production']
});

var ResHelper = require('./ResHelper.js');


var Globals = {
    io : {},
    development: env === 'development',
    production : env === 'production',
    saveErrorCallback : function(err){
        if (err){
            console.log(err);
        }
    },
    fetchErrorCallback : function(res){
        return function(err){
            ResHelper.sendError(res, error);
        };
    },
    RETRIES_ON_DUPLICATE_CONFIRMATION_CODE: 5,
    genders : ["male", "female"],
    colleges : ["nyu", "baruch"],
    providers : ["facebook", "userpw"],
    soireeTypes : ["lunch", "dinner", "drinks", "blind date", "TEST"],
    initialChargePerSoireeType : {"lunch" : 99, "dinner" : 99, "drinks" : 99, "blind date" : 99, "TEST" : 99},
    availableTimesPerSoireeType : {
        "lunch" : ["12:00 PM", "12:30 PM", "1:00 PM", "1:30 PM", "2:00 PM", "2:30 PM"],
        "dinner" : ["6:00 PM", "6:30 PM", "7:00 PM", "7:30 PM"],
        "drinks" : ["9:00 PM", "9:30 PM", "10:00 PM", "10:30 PM"],
        "blind date" : ["7:00 PM", "8:00 PM", "9:00 PM"],
        "TEST" : ["1:00 PM", "2:00 PM"]
    },
    numUsersMaxPerSoireeType : {"lunch" : 4, "dinner" : 4, "drinks" : 6, "blind date" : 2, "TEST" : 3},

    collegeTimezones : {
        "nyu" : "est", "baruch" : "est"
    }

};


module.exports = Globals;
