// app/db/Image.js
/**
 * Created by shadygabal on 1/17/16.
 */
var mongoose = require('./mongoose_connect.js');
var Schema = mongoose.Schema;
var ObjectId = Schema.Types.ObjectId;
var UserVerification = require('./UserVerification.js');

var imageSchema = new Schema({
        data: Buffer,
        contentType: String,
        fileName : {type: String, required: true},
        directory : {type: String, required: true},
        path : {type: String, index: true, sparse: true},
        adminsOnly : {type: Boolean, default: false},
        _userVerification : {type: ObjectId, ref:"UserVerification"},
        _user : {type: ObjectId, ref:"User", index: true, sparse: true}
    },
    { timestamps: { createdAt: 'dateCreated', updatedAt: 'dateUpdated' } }
);

function createPath(directory, fileName){
    //console.log("setting path...");
    if (directory && fileName) {

        if (directory.charAt(0) !== '/') {
            directory = '/' + directory;
        }
        if (directory.charAt(directory.length - 1) !== '/') {
            directory = directory + '/';
        }

        var path = directory + fileName;
        //console.log("set path to: " + path);
        return path;

    }
    return null;
}

imageSchema.statics.createPath = function(directory, fileName){
   return createPath(directory, fileName);
};


imageSchema.pre('validate', function(next){
    if (this.directory && this.directory.charAt(this.directory.length - 1) !== '/') {
        this.directory = this.directory + '/';
    }
    if (this.directory && this.directory.charAt(0) !== '/') {
        this.directory = '/' + this.directory;
    }
    if (!this.directory){
        this.directory = "/images/";
    }
    if (!this.path){
        this.path = createPath(this.directory, this.fileName);
    }
    next();
});

//imageSchema.pre('save', function(next){
//    console.log("image pre save");
//    next();
//});

//imageSchema.post('init', function(doc){
//    console.log("post init");
//    //doc.setPath();
//});

imageSchema.virtual('url').get(function() {
    return "http://www.experiencesoiree.com" + this.path;
});

var deepPopulate = require('mongoose-deep-populate')(mongoose);
var options = {};
imageSchema.plugin(deepPopulate, options);

module.exports = mongoose.model('Image', imageSchema);
