// app/db/Notification.js
/**
 * Created by shadygabal on 1/18/16.
 */
var mongoose = require('./mongoose_connect.js');
var Schema = mongoose.Schema;
var ObjectId = Schema.Types.ObjectId;

/* Other Models */
var User = require('./User.js');

/* Packages */
var shortid = require('shortid');

/* Helper */
var helpersFolderLocation = "../helpers/";
var DateHelper = require('app/helpers/DateHelper.js');
var ResHelper = require('app/helpers/ResHelper.js');
var CreditCardHelper = require('app/helpers/CreditCardHelper.js');
var LocationHelper = require('app/helpers/LocationHelper.js');
var PushNotificationHelper = require('app/helpers/PushNotificationHelper.js');
var MongooseHelper = require('app/helpers/MongooseHelper.js');
var Globals = require('app/helpers/Globals.js');

var notificationTypes = ["commented", "liked", "soireeCancelled", "soireeReminder", "soireeStarting", "soireeConfirmed", "userAccepted", "userRejected"];

var notificationTypesObj = {
    "commented" : "commented",
    "liked" : "liked",
    "soireeReminder" : "soireeReminder",
    "soireeStarting" : "soireeStarting",
    "soireeCancelled" : "soireeCancelled",
    "soireeConfirmed" : "soireeConfirmed",
    "userAccepted" : "userAccepted",
    "userRejected" : "userRejected"
};

var notificationSchema = new Schema({
        read: {type: Boolean, default: false},
        users: [{user : String, firstName: String, lastName: String}],
        title : String,
        bodySuffix : {type: String, required: true},
        postId : {type: String},
        imageUrl : {type: String},
        notificationType : {type: String, enum: notificationTypes, required: true},
        _user : {type: ObjectId, ref: "User"},
        _id : {type: String},
        date : {type: Date, default: new Date()}
});

notificationSchema.statics.notificationTypes = notificationTypesObj;

//for creating notifications when user comments on a post

notificationSchema.statics.createCommentedOnPostNotifications = function(userThatCommented, upPost, upComment){//up = un populated
    var Notification = this;
    console.log("createCommentedOnPostNotification()");

    //first make sure you have populated all required fields for both the post and the comment
    upPost.deepPopulate("_user _comments._user", function(err2, post){
        if (err2 || !post){ return console.log("Error fetching post: " + err2); }

        upComment.deepPopulate("_user", function(err3, comment){
            if (err3 || !comment){ return console.log("Error fetching comment: " + err3); }

            //if person that commented is NOT the post author, send a notification to the post author
            if (!MongooseHelper.equalsPopulated(userThatCommented, post._user)) {
                var body = ' commented on your post "' + post.text + '"';
                Notification.pushCommunityNotification(body, post._user, comment._user, post, Notification.notificationTypes.commented);
            }
            //go through comments, pick out all unique users and notify them
            //TODO: OPTIMIZE
            var commentUsersNotified = [];
            for (var i = 0; i < post._comments; i++){
                var postComment = post._comments[i];
                var commentUser = postComment._user;
                //if havent notified user before
                if (commentUsersNotified.indexOf(commentUser._id) === -1 && !MongooseHelper.equalsPopulated(commentUser, userThatCommented) && !MongooseHelper.equalsPopulated(commentUser, post._user)){
                    var bodySuffix = " commented on a post you commented on \"" + post.text + "\"";

                    Notification.pushCommunityNotification(bodySuffix, commentUser, userThatCommented, post, Notification.notificationTypes.commented);
                    commentUsersNotified.push(commentUser._id);
                }
            }

        });
    });
    //});
};


//if notification already exists, modify it to add this on. if not, create one. then push
notificationSchema.statics.pushCommunityNotification = function(bodySuffix, notificationsUser, causingUser,  post, type){

    var Notification = this;
    console.log("pushCommunityNotification() with notifications user: " + notificationsUser.fullName);

    var idToMatch = generateId(notificationsUser, post.postId, type);
    var index = notificationsUser._notifications.indexOf(idToMatch);

    if (index != -1){
        var notificationId = notificationsUser._notifications[index];

        Notification.findOne({_id : notificationId}).exec(function(err, notification){
            if (err){
                console.log("addToNotification() error fetching notification with id: " + notificationId);
            }
            else if (!notification){
                console.log("addToNotification() no notification found with id: " + notificationId);
                Notification.createCommunityNotification(bodySuffix, notificationsUser, causingUser, post, type);
            }
            else{
                var newUser = {user : causingUser.id, firstName: causingUser.firstName, lastName: causingUser.lastName};

                var filterOutExistingUsers = function (userObj){
                    return userObj.user !== newUser.user;
                };

                notification.users = notification.users.filter(filterOutExistingUsers);
                console.log("Filtered out notification.users to: " + notification.users);
                notification.users.push(newUser);
                notification.title = causingUser.fullName;
                notification.imageUrl = causingUser.profilePictureUrl;
                notification.read = false;
                notification.date = new Date();
                notification.save(Globals.saveErrorCallback);
                //save and push notification to user
                notificationsUser.addNotification(notification);

            }
        });
    }
    else{
       console.log("addToNotification(): idToMatch " + idToMatch + " was not found in notificationsUser._notifications: " + notificationsUser._notifications + ". Creating...");
        Notification.createCommunityNotification(bodySuffix, notificationsUser, causingUser, post, type);
    }

};

notificationSchema.statics.createCommunityNotification = function(bodySuffix, notificationsUser, causingUser, post, type){
    var Notification = this;

    var firstUser = {user: causingUser.id, firstName: causingUser.firstName, lastName: causingUser.lastName};
    //console.log("firstUser: " + JSON.stringify(firstUser));
    var fullName = causingUser.fullName;

    var newNotification = new Notification({
        bodySuffix : bodySuffix,
        _user : notificationsUser._id,
        postId : post.postId,
        imageUrl : causingUser.profilePictureUrl,
        notificationType : type,
        _id : generateId(notificationsUser, post.postId, type),
        users: [firstUser],
        title : fullName
    });

    newNotification.save(function(err, notif){
        if (err){
            console.log(err);
        }
        else{
            console.log("Created Notification: " + newNotification.body + " for: " + notificationsUser.firstName);
            notificationsUser.addNotification(notif);

        }
    });
};

notificationSchema.statics.createUserVerifiedNotification = function(user){
    if (!user){
        console.log("null user passed to createUserVerifiedNotification()");
        return;
    }
    var Notification = this;
    var message = "Congratulations! You've been accepted as a member of " + Globals.SOIREE + "! You can now join " + Globals.SOIREE_LOWERCASE + "s and post in community!";

    var newNotification = new Notification({
       bodySuffix: message,
        _user : user._id,
        notificationType : Notification.notificationTypes.userAccepted,
        _id : generateId(user, "" + Date.now(), Notification.notificationTypes.userAccepted),
        title: "Verification accepted!"
    });

    newNotification.save(function(err, notif){
        if (err){
            console.log(err);
        }
        else{
            console.log("Created Notification: " + newNotification.body + " for: " + user.firstName);
            user.addNotification(notif);
        }

    });
};

notificationSchema.statics.createUserRejectedNotification = function(user, reason){
    if (!user){
        console.log("null user passed to createUserVerifiedNotification()");
        return;
    }
    if (!reason){
        console.log("null reason passed to createUserVerifiedNotification()");
        reason = "the images were unreadable / the face or name were blocked out";
    }
    else{
        reason = reason.toLowerCase();
    }

    var Notification = this;
    var message = "There was a problem with your user verification. It was rejected because " + reason;

    var newNotification = new Notification({
        bodySuffix: message,
        _user : user._id,
        notificationType : Notification.notificationTypes.userAccepted,
        _id : generateId(user, "" + Date.now(), Notification.notificationTypes.userRejected),
        title: "Verification rejected :("
    });

    newNotification.save(function(err, notif){
        if (err){
            console.log(err);
        }
        else{
            console.log("Created Notification: " + newNotification.body + " for: " + user.firstName);
            user.addNotification(notif);
        }

    });
};

notificationSchema.statics.notificationsForUser = function(user, successCallback, errorCallback) {
};

notificationSchema.statics.jsonArrayFromArray = function(_notifications) {
    var notifications = [];
    for (var i = _notifications.length-1; i >= 0; i--){
        var notification = _notifications[i];
        notifications.push(notification.jsonObject());
    }
    return notifications;
};

notificationSchema.statics.createSoireeCancelledNotification = function(soiree, notificationsUser){
    var Notification = this;

    var message = "Your " + soiree.soireeType + " " + Globals.SOIREE_LOWERCASE + " didn't have enough people, so we had to cancel it."
    "Don't worry, you weren't charged for it.";

    var newNotification = new Notification({
        bodySuffix : message,
        _user : notificationsUser._id,
        notificationType : Notification.notificationTypes.soireeCancelled,
        _id : generateId(notificationsUser, soiree.soireeId, Notification.notificationTypes.soireeCancelled),
        title : 'Cancelled ' + Globals.SOIREE_LOWERCASE
    });

    newNotification.save(Globals.saveErrorCallback);
    notificationsUser.addNotification(newNotification);

    console.log("Created Notification: " + newNotification.body + " for: " + notificationsUser.firstName);
    return newNotification;

};

notificationSchema.statics.createSoireeConfirmedNotification = function(soiree, notificationsUser){
    var Notification = this;

    var message = "Enough people have joined your " + soiree.soireeType + " " + Globals.SOIREE_LOWERCASE + ", so it's officially on! See you there."

    var newNotification = new Notification({
        bodySuffix : message,
        _user : notificationsUser._id,
        notificationType : Notification.notificationTypes.soireeConfirmed,
        _id : generateId(notificationsUser, soiree.soireeId, Notification.notificationTypes.soireeCancelled),
        title : 'Confirmed ' + Globals.SOIREE_LOWERCASE
    });

    newNotification.save(Globals.saveErrorCallback);
    notificationsUser.addNotification(newNotification);

    console.log("Created Notification: " + newNotification.body + " for: " + notificationsUser.firstName);
    return newNotification;

};

notificationSchema.statics.createSoireeReminderNotification = function(soiree, notificationsUser, mins) {
    var Notification = this;

    var message = "Hey boo. Don't forget that your " + soiree.soireeType + " " + Globals.SOIREE_LOWERCASE + " will start in " + mins + " minutes. See you there. xoxo";

    var newNotification = new Notification({
        bodySuffix : message,
        _user : notificationsUser._id,
        notificationType : Notification.notificationTypes.soireeReminder,
        _id : generateId(notificationsUser, soiree.soireeId, Notification.notificationTypes.soireeReminder),
        title : Globals.SOIREE + " reminder"
    });

    newNotification.save(function(err){
       if (err){
           console.log(err);
       }
        else{
           notificationsUser.addNotification(newNotification);
           console.log("Created Notification: " + newNotification.body + " for: " + notificationsUser.firstName);
       }
    });
};

notificationSchema.statics.createSoireeStartingNotification = function(soiree, notificationsUser) {
    var Notification = this;

    //var message = "Hey boo. Don't forget that your " + soiree.soireeType + " " + Globals.SOIREE_LOWERCASE + " will start in " + mins + " minutes. See you there. xoxo";

    var message = "Your " + soiree.soireeType + " " + Globals.SOIREE_LOWERCASE + " is about to start! Open up " + Globals.SOIREE + " to get started.";

    var newNotification = new Notification({
        bodySuffix : message,
        _user : notificationsUser._id,
        notificationType : Notification.notificationTypes.soireeReminder,
        _id : generateId(notificationsUser, soiree.soireeId, Notification.notificationTypes.soireeStarting),
        title : Globals.SOIREE + " starting"
    });

    newNotification.save(function(err){
        if (err){
            console.log(err);
        }
        else{
            notificationsUser.addNotification(newNotification);
            console.log("Created Notification: " + newNotification.body + " for: " + notificationsUser.firstName);
        }
    });
};


notificationSchema.methods.jsonObject = function(){
    var obj = {
        "read" : this.read,
        "body" : this.body,
        "notificationType" : this.notificationType,
        "notificationTitle" : this.notificationTitle,
        "notificationId" : this._id,
        "date" : this.date
    };

    if (this.postId){
        obj.postId = this.postId;
    }
    //if (this.lastUserCommentedName){
    //    obj.lastUserCommentedName = this.lastUserCommentedName;
    //}
    if (this.imageUrl){
        obj.imageUrl = this.imageUrl;
    }

    return obj;
};

notificationSchema.virtual('notificationTitle').get(function () {
    if (this.notificationType === "soireeCancelled"){
        return Globals.SOIREE + " cancelled";
    }
    else return this.lastUserCommentedName;
});

notificationSchema.virtual('lastUserCommentedName').get(function () {
    if (this.users.length > 0){
        var lastUserObj = this.users[this.users.length-1];
        return lastUserObj.firstName + " " + lastUserObj.lastName;
    }
    else return "Notification";
});

notificationSchema.virtual('body').get(function () {
    if (!this.users || this.users.length === 0){
        return this.bodySuffix;
    };

    console.log("this.users:");
    for (var k = 0; k < this.users.length; k++){
        console.log(this.users[k]);
    }

    var numNames = this.users.length;

    if (numNames == 0)
        return this.bodySuffix;

    var firstName = this.users[0].firstName;

    if (numNames === 1){
        return firstName + this.bodySuffix;
    }

    var secondName = this.users[1].firstName;

    if (numNames === 2){
        return firstName + " and " + secondName + this.bodySuffix;
    }

    numNames -= 2;
    var others = (numNames === 1 ? " other" : " others");

    return firstName + ", " + secondName + ", and " + numNames + others + this.bodySuffix;
});

function generateId(user, idSuffix, type){
//USER SHOULD ONLY EVER BE THE USER THAT IS RECEIVING THE NOTIFICATION
    var id = user.id + "_" + idSuffix + "_" + type;
    console.log("Notification.generateId(): " + id);
    return id;
}


//notificationSchema.pre('save', function(next){
//    if (!this.notificationId){
//        IdGeneratorHelper.generateUniqueId("Notification", "notificationId", this, next, 15);
//    }
//    else next();
//});


var deepPopulate = require('mongoose-deep-populate')(mongoose);
var options = {};
notificationSchema.plugin(deepPopulate, options);

module.exports = mongoose.model('Notification', notificationSchema);
