// app/db/Soiree.js
/* Setup */
var mongoose = require('./mongoose_connect.js');
var Schema = mongoose.Schema;
var ObjectId = Schema.Types.ObjectId;

/* Other Models */


/* Packages */
var shortid = require('shortid');
var _ = require("underscore");

/* Helper */
var helpersFolderLocation = "../helpers/";
var Globals = require('app/helpers/Globals.js');
var DateHelper = require('app/helpers/DateHelper.js');
var ResHelper = require('app/helpers/ResHelper.js');
var CreditCardHelper = require('app/helpers/CreditCardHelper.js');
var ArrayHelper = require('app/helpers/ArrayHelper.js');
var LocationHelper = require('app/helpers/LocationHelper.js');
var PushNotificationHelper = require('app/helpers/PushNotificationHelper.js');



/* Defaults */
var numUsersMaxPerSoireeType = { "lunch" : 4, "dinner" : 4, "drinks" : 6, "blind date" : 2 };

var AVERAGE_SOIREE_DURATION = 45;
var MIN_TIME_BEFORE_CAN_CREATE_SOIREE = 0;//1.5 * 60 * 60 * 1000
/* Error Codes */
var ErrorCodes = require('app/helpers/ErrorCodes.js');

var timeCountsPerCollege = {};

var soireeSchema = new Schema({
		soireeType : {type: String, required: true, enum: Globals.soireeTypes},
		soireeDescription : {type: String},
		numUsersMax: {type : Number, required: true},
		numUsersMin: {type : Number, required: true},
		scheduledStartTimeIdentifier : {type: String},
		scheduledEndTimeIdentifier : {type: String},
		soireeId: {type: String, index: true, default: shortid.generate},
		initialCharge: {type: Number, required: [true, "Forgot to include how much soiree will cost"]}, //in cents
		date: {type : Date, required: [true, "A date for the Soiree is required"]},
		_usersAttending : [{type : ObjectId, ref : "User"}],
		_usersUncharged : [{type : ObjectId, ref : "User"}],
		_business: {type: ObjectId, ref:"Business", required :[true, "A business that will host is required to create this Soiree"]},
		expired: {type: Boolean, default: false},
		location: {
			type: {type: String},
			coordinates: []
		},
		photoIndexIdentifier : {type: Number, default: generatePhotoIndexIdentifier},
		started : {type: Boolean, default: false},
		college : {type: String, enum: Globals.colleges},
		ended : {type: Boolean, default: false},
		open : {type: Boolean, default: false},
		duration : {type: Number, default: AVERAGE_SOIREE_DURATION},
		_unchargedReservations : [{type: ObjectId, ref: "SoireeReservation"}],
		_chargedReservations : [{type: ObjectId, ref: "SoireeReservation"}],
		_host : {type: ObjectId, ref: "SoireeHost"}


	},
	{ timestamps: { createdAt: 'dateCreated', updatedAt: 'dateUpdated' } }
);

soireeSchema.index({location: '2dsphere'});

soireeSchema.statics.SOIREE = "Soirée";
soireeSchema.statics.SOIREE_LOWERCASE = "soirée";
soireeSchema.statics.AVERAGE_SOIREE_DURATION = AVERAGE_SOIREE_DURATION;
soireeSchema.statics.AVERAGE_SOIREE_DURATION_IN_MS = AVERAGE_SOIREE_DURATION * 60 * 1000;


function generatePhotoIndexIdentifier(){
	var rand = parseInt(Math.random() * 1000);
	return rand;
};


/* Static Methods */
//soireeSchema.statics.errorCodes = function() {
//	return this.errorCodes;
//}

soireeSchema.statics.createScheduledTimeIdentifier = function(date){
	if (!date){
		date = new Date();
	}
	if (date.constructor != Date){
		date = new Date(date);
	}

	var year = date.getFullYear();
	var month = date.getMonth() + 1;
	var day = date.getDate();

	var hours = date.getHours();
	var mins = date.getMinutes();
	//round mins to nearest 10
	if (date.getSeconds() >= 30)
		mins = mins + 1;
	mins -= (mins % 10);

	return "" + year + "." + month + "." + day + "." + hours + "." + mins;
};

soireeSchema.statics.createScheduledTimeIdentifierPast = function(mins){
	return this.createScheduledTimeIdentifier(Date.now() - (mins * 60 * 1000));
};

soireeSchema.statics.createScheduledTimeIdentifierFuture = function(mins){
	return this.createScheduledTimeIdentifier(Date.now() + (mins * 60 * 1000));
};

soireeSchema.statics.findNextSoiree = function(user, idsToIgnore, successCallback, errorCallback){
	if (!idsToIgnore) idsToIgnore = [];
	var today = new Date();
	var college = user.college;
	console.log('college: ' + college);

	this.find({college: college, soireeId : {"$nin" : idsToIgnore}, date: {"$gte" : today}}).sort({"date" : 1}).limit(1).exec(function(err, soirees){
		console.log(soirees);

		if (err || !soirees){
			console.log("error finding next soiree: " + err);
			errorCallback(ErrorCodes.MongoError);
		}
		else{
			console.log("found soirees: " + soirees);
			if (soirees.length === 0){
				successCallback(null);
			}
			else successCallback(soirees[0]);
		}
	});

};
//soireeSchema.statics.findSoireesWithScheduledTimeIdentifier = function(scheduledTimeIdentifier, successCallback, errorCallback){
//	Soiree.find({"scheduledTimeIdentifier" : scheduledTimeIdentifier}).populate("_business").exec(function(err, soirees){
//		if (err){
//			errorCallback(err);
//		}
//		else{
//			successCallback(soirees);
//		}
//	});
//};

//soireeSchema.methods.joinUsers = function(users, successCallback, errorCallback){
//
//};

soireeSchema.statics.createSoireesWithJobs = function(ssJobs, soireeType, callback){
	console.log('creating soirees with jobs: ' + ssJobs + ' for type: ' + soireeType);

	var Soiree = this;

	var numUsersMax = Globals.numUsersMaxPerSoireeType[soireeType];
	var numUsersMin = Globals.numUsersMinPerSoireeType[soireeType];
	var leftovers = ssJobs.length % numUsersMin;

	var numSoireesCanCreate = parseInt(ssJobs.length / numUsersMin);
	console.log("numSoireesCanCreate: " + numSoireesCanCreate);
	console.log('ssJobs: ' + ssJobs);

	if (numSoireesCanCreate == 0){
		console.log("cant create any soirees. returning");
		return;
	}

	var college = ssJobs[0].college;

	var soireesNumUsersArray =[];
	for (var j = 0; j < numSoireesCanCreate; j++){
		var num = numUsersMin;
		if (numUsersMin < numUsersMax && leftovers > 0){
			leftovers--;
			num++;
		}
		soireesNumUsersArray[j] = num;
	}
	console.log('soireesNumUsersArray: ' + soireesNumUsersArray);

	var numSoireesCreated = 0;

	var runBatchSSJobs = function(currRound){
		var users = [];
		var num = soireesNumUsersArray[currRound];

		//for (var i = numSoireesCreated; i < numSoireesCreated + num; i++){
		//	var ssJob = ssJobs[i];
		//	users.push(ssJob._user);
		//}
		var startIndex = numSoireesCreated;
		var endIndex = numSoireesCreated + num;
		numSoireesCreated += num;

		console.log("creating soiree for users: " + users);

		Soiree.createSoireeWithType(soireeType, college, function(soiree){
			console.log("created soiree " + soiree + ' . users joining...');
			var currIndex = startIndex;
			var joinUserSuccessCallback, joinUserErrorCallback;

			var joinUser = function(index) {
				var ssJob = ssJobs[index];
				var user = ssJob._user;
				console.log("user: " + user._id + " joining...");
				soiree.join(user, joinUserSuccessCallback, joinUserErrorCallback);
			};

			var performNext = function(){
				currIndex++;
				if (currIndex < endIndex){
					joinUser(currIndex);
				}
				else console.log("done");
			};

			joinUserSuccessCallback = function() {
				console.log("success");
				performNext();
			};

			joinUserErrorCallback = function(err){
				console.log("error: " + err);
				performNext();
			};

			joinUser(currIndex);


		}, function(err){
			console.log("error creating soiree");

		});
	};


	for (var a = 0; a < numSoireesCanCreate; a++){
		runBatchSSJobs(a);
	}

};

soireeSchema.statics.setSoireeDefaults = function(soiree, successCallback, errorCallback, options){
	var Soiree = this;
	var Business = mongoose.model("Business");

	if (!options) options = {};

	var initialCharge = options.initialCharge;
	var numUsersMax = options.numUsersMax;
	var numUsersMin = options.numUsersMin;

	if (!numUsersMax){
		numUsersMax = Globals.numUsersMaxPerSoireeType[soiree.soireeType];
	}
	if (!numUsersMin){
		numUsersMin = Globals.numUsersMinPerSoireeType[soiree.soireeType];
	}
	if (!initialCharge){
		initialCharge = Globals.initialChargePerSoireeType[soiree.soireeType];
	}

	soiree.initialCharge = initialCharge;
	soiree.numUsersMax = numUsersMax;
	soiree.numUsersMin = numUsersMin;

	if (soiree.date && soiree.college) {
		var hoursMinsString = DateHelper.timeStringFromDate(soiree.date);

		if (soiree.soireeType != "TEST"){
			timeCountsPerCollege[soiree.college][soiree.soireeType][hoursMinsString]++;
		}
	}

	if (options["business"]){
		Soiree.createSoireeWithBusiness(soiree, options.business, successCallback, errorCallback);
	}
	else{
		Business.nextBusinessToHostSoiree(soiree.college, function(business){
			if (!business){ return errorCallback(ErrorCodes.NotFound);}

			saveSoireeWithBusiness(soiree, business, successCallback, errorCallback);
		}, function(err){
			errorCallback(err);
		});
	}
};

function saveSoireeWithBusiness(soiree, business, successCallback, errorCallback) {
	//if (_.difference(soiree.colleges, business.colleges).length > 0){
	//	return errorCallback(ErrorCodes.InvalidInput);
	//}

	var Soiree = mongoose.model("Soiree");

	if (business.colleges.indexOf(soiree.college) === -1){
		return errorCallback(ErrorCodes.InvalidInput);
	}

	var newSoiree = new Soiree(soiree);

	newSoiree._business = business._id;
	newSoiree.location = business.location;
	newSoiree._usersAttending = [];


	newSoiree.save(function(err, savedSoiree){
		if (err){
			console.log(err);
			errorCallback(ErrorCodes.ErrorSaving);
		}
		else{
			successCallback(savedSoiree);
		}
	});
};

soireeSchema.statics.findSoireesForBusiness = function(business, options, successCallback, errorCallback){
	if (!business)
		return errorCallback(ErrorCodes.MissingData);

	_.defaults(options, {showExpired: false, sameDay: false});

	this.find({"_business" : business._id, expired: options.showExpired}).deepPopulate("_usersAttending").exec(function(err, soirees){
		if (err){
			console.log(err);
			errorCallback(ErrorCodes.ErrorQuerying);
		}
		else{
			successCallback(soirees);
		}
	});
};

soireeSchema.statics.createAppropriateSoiree = function(){
	var date = new Date();
	var hour = date.getHours();

};

soireeSchema.statics.createSoireeWithType = function(soireeType, college, successCallback, errorCallback, options) {
	if (soireeType === "Lunch"){
		this.createLunch(college, successCallback, errorCallback);
	}
	else if (soireeType === 'Dinner'){
		this.createDinner(college, successCallback, errorCallback);
	}
	else if (soireeType === 'Drinks'){
		this.createDrinks(college, successCallback, errorCallback);
	}
	else if (soireeType === 'Blind Date'){
		this.createBlindDate(college, successCallback, errorCallback);
	}
	else if (soireeType === 'TEST'){
		this.createTest(college, successCallback, errorCallback);
	}
	else return errorCallback(ErrorCodes.InvalidInput);
};


soireeSchema.statics.fillTimeCountsPerCollege = function(){
	for (var i = 0; i < Globals.colleges.length; i++){
		var college = Globals.colleges[i];

		for (var j = 0; j < Globals.soireeTypes.length; j++){
			var soireeType = Globals.soireeTypes[j];

			this.setCorrespondingTimeCount(soireeType, college, function(counts, _soireeType, _college){
				if (!timeCountsPerCollege[_college]) timeCountsPerCollege[_college] = {};
				if (!timeCountsPerCollege[_college][_soireeType]) timeCountsPerCollege[_college][_soireeType] = {};

				timeCountsPerCollege[_college][_soireeType] = counts;
			});


		};

	}
}

soireeSchema.statics.setCorrespondingTimeCount = function(soireeType, college, callback){
	this.countsForSoireeTypeAtTimes(soireeType, college, function(counts){
		callback(counts, soireeType, college);
	});
};

soireeSchema.statics.countsForSoireeTypeAtTimes = function(soireeType, college, callback) {
	var Soiree = this;

	var availableTimes = Globals.availableTimesPerSoireeType[soireeType];

	var numToReturn = availableTimes.length;
	var numReturned = 0;
	var numErred = 0;
	var ans = {};

	var countCallback = function(err, count, atTime){

		numReturned++;

		if (err){
			numErred++;
			console.log(err);
			if (numErred === numToReturn){
				return callback(null);
			}
		}
		else{
			ans[atTime] = count;
		}
		//console.log("count: " + count + " for " + atTime);
		//console.log(ans);

		if (numReturned === numToReturn){
			//console.log("returning");
			callback(ans);
		}
	};

	for (var i = 0; i < availableTimes.length; i++){
		var timeString = availableTimes[i];

		this.performCountQuery(soireeType, college, timeString, countCallback);

	}
};

soireeSchema.statics.performCountQuery = function(soireeType, college, atTime, callback){
	var Soiree = this;
	var date = DateHelper.dateFromTime(atTime);

	Soiree.count({soireeType : soireeType, date : date, college : college}, function(err, count){
		callback(err, count, atTime);
	});

}



soireeSchema.statics.dateForSoireeType = function(soireeType){
	var availableTimes = Globals.availableTimesPerSoireeType[soireeType];
	if (!availableTimes || availableTimes.length == 0){
		return null;
	}
};

soireeSchema.statics.createLunch = function(college, successCallback, errorCallback, options) {
	//var date = new Date(todaysDate.getTime() + (7 * 24 * 60 * 60 * 1000));
	var Soiree = this;

	if (!college) return errorCallback();

	var soireeType = "Lunch";

	determineDateForSoireeType(soireeType, college, function(date){
		if (!date) return errorCallback(ErrorCodes.NoAvailableDate);

		var soiree = new Soiree({
			soireeType: soireeType,
			date: date,
			college : college
		});

		Soiree.setSoireeDefaults(soiree, successCallback, errorCallback, options);
	});

};

soireeSchema.statics.createDinner = function(college, successCallback, errorCallback, options) {
	if (!college) return errorCallback();

	var Soiree = this;

	var soireeType = "Dinner";

	determineDateForSoireeType(soireeType, college, function(date){
		if (!date) return errorCallback(ErrorCodes.NoAvailableDate);

		var soiree = new Soiree({
			soireeType: soireeType,
			date: date,
			college : college
		});

		Soiree.setSoireeDefaults(soiree, successCallback, errorCallback, options);
	});

};

soireeSchema.statics.createTest = function(college, successCallback, errorCallback, options) {
	if (!college) return errorCallback();

	var Soiree = this;

	var today = new Date();
	var mins = today.getMinutes();
	var newMins = mins + (10 - (mins % 10));

	var newDate = new Date();
	newDate.setMinutes(newMins);


	var soireeType = "TEST";

	//determineDateForSoireeType(soireeType, college, function(date){
	//	if (!date) return errorCallback(ErrorCodes.NoAvailableDate);

		var soiree = new Soiree({
			soireeType: soireeType,
			date: newDate,
			college : college
		});

		Soiree.setSoireeDefaults(soiree, successCallback, errorCallback, options);
	//});

};

soireeSchema.statics.createDrinks = function(college, successCallback, errorCallback, options) {
	if (!college) return errorCallback();
	var Soiree = this;

	var soireeType = "Drinks";
	determineDateForSoireeType(soireeType, college, function(date){
		if (!date) return errorCallback(ErrorCodes.NoAvailableDate);

		var soiree = new Soiree({
			soireeType: soireeType,
			date: date,
			college : college
		});

		Soiree.setSoireeDefaults(soiree, successCallback, errorCallback, options);
	});

};

soireeSchema.statics.createBlindDate = function(college, successCallback, errorCallback, options) {
	//var date = new Date(todaysDate.getTime() + (7 * 24 * 60 * 60 * 1000));
	if (!college) return errorCallback();

	var Soiree = this;

	var soireeType = "Blind Date";
	determineDateForSoireeType(soireeType, college, function(date){
		if (!date) return errorCallback(ErrorCodes.NoAvailableDate);

		var soiree = new Soiree({
			soireeType: soireeType,
			date: date,
			college : college
		});

		Soiree.setSoireeDefaults(soiree, successCallback, errorCallback, options);
	});

};

soireeSchema.statics.findSoirees = function(req, user, successCallback, errorCallback){

	var constraints = {};

	if (req.body.user){
		var longitude = req.body.user.longitude;
		var latitude = req.body.user.latitude;
		var coors = LocationHelper.createPoint(longitude, latitude);

		constraints.location = { $near : coors } ;

	}


	var numSoireesToFetch = 10;

	var idsToIgnore = req.body.currentSoireesIds;

	//var constraints = { location: { $near : coors }, "college" : user.college };

	if (idsToIgnore && idsToIgnore.length > 0){
		console.log("Ignoring soirees with ids in: " + idsToIgnore);
		constraints["soireeId"] = {'$nin' : idsToIgnore};
	}

	this.find(constraints).deepPopulate("_business _usersAttending.college").limit(numSoireesToFetch).exec(function(err, soirees){
		if (err){
			errorCallback(err);
		}
		else{
			successCallback(soirees);
		}
	});
};


//soireeSchema.statics.timeAtStringFromDate = function(date){
//	//var day = (24 + i) % 30;
//	//var date = new Date(2015, (10 - 1), day, 18, 30, 0);
//	var todaysDate = new Date();
//	var hour = date.getHours();
//	var minutes = date.getMinutes();
//
//	if (minutes < 10){
//		minutes = "0" + minutes.toString();
//	}
//
//	var amPm = "AM";
//	var when = "Today";
//
//	if (!DateHelper.isSameDay(todaysDate, date)){
//		if (DateHelper.isNextDay(todaysDate, date)){
//			when = "Tomorrow";
//		}
//		else when = DateHelper.dayFromDayNumber(date.getDay());
//	}
//	else if (hour > 18){
//		when = "Tonight";
//	}
//
//	if (hour > 12) {
//		hour -= 12;
//		amPm = "PM";
//	}
//	var timeAtString = when + " at " + hour + ":" + minutes + " " + amPm;
//	return timeAtString;
//};


//soireeSchema.statics.soireeTypes = function(){
//	return soireeTypes;
//};


soireeSchema.statics.findBySoireeId = function(soireeId, successCallback, errorCallback){
	this.findOne({soireeId : soireeId}).populate('_business').exec(function(err, soiree){
		if (err){
			return errorCallback(err);
		}
		else if(!soiree){
			return errorCallback();
		}
		else{
			successCallback(soiree);
		}
	});
};

soireeSchema.statics.joinSoireeWithId = function(soireeId, user, successCallback, errorCallback){
	this.findBySoireeId(soireeId, function(soiree){
		soiree.join(user, successCallback, errorCallback);
	}, function(err){
		errorCallback(ErrorCodes.SoireeError);
	});
};

/* Methods */

soireeSchema.methods.remind = function(mins) {
	console.log("Reminding users of soiree " + this.soireeType + " " + this.scheduledStartTimeIdentifier + " with users attending: " + this._usersAttending + " ...");

	for (var i = 0; i < this._usersAttending.length; i++){
		var user = this._usersAttending[i];
		console.log("Sending push notification to " + user.firstName);

		var message = "Hey boo. Don't forget that your " + this.soireeType + " " + this.SOIREE + " will start in " + mins + " minutes. See you there. xoxo";
		PushNotificationHelper.sendPushNotification(user, message);
	}
};

soireeSchema.methods.start = function(){
	if (!this.open){
		this.openToUsers();
	}
	this.deepPopulate("_usersAttending", function(err, soiree){
		if (err){
			console.log(err);
		}

		console.log("Starting soiree " + soiree.soireeType + " " + soiree.scheduledStartTimeIdentifier + " with users attending: " + soiree._usersAttending + " ...");

		if (!soiree._host){
			var SoireeHost = mongoose.model('SoireeHost');

			var host = new SoireeHost({
				_soiree : soiree._id,
				roomId : soiree.soireeId
			});
			soiree._host = host._id;
			host.save(Globals.saveErrorCallback);
		}

		alertUsersThatSoireeStarted(soiree);

		soiree.started = true;
		soiree.inProgress = true;

		soiree.save(Globals.saveErrorCallback);
	});
};


soireeSchema.methods.end = function() {
	console.log("Ending soiree " + this.soireeType + " " + this.scheduledStartTimeIdentifier + " with users attending: " + this._usersAttending + " ...");

	var soiree = this;

	this.deepPopulate("_usersAttending", function(err){
		if (err){
			console.log("Error ending soiree: " + err);
		}
		else{
			for (var i = 0; i < soiree._usersAttending.length; i++) {
				var user = soire._usersAttending[i];
				user.endedSoiree(soiree);
			}
		}
		soiree.ended = true;
		soiree.inProgress = false;
		soiree.save(Globals.saveErrorCallback);
	});
};

soireeSchema.methods.openToUsers = function() {
	this.open = true;
	this.save(function(err){
		if (err){
			console.log("Error opening soiree: " + err);
		}
	});
};

soireeSchema.methods.cancelSoireeIfNecessary = function(){
	if (!this.reachedNumUsersMin){
		//cancel soiree
	}
};

soireeSchema.methods.hasUserAlreadyJoined = function(user){
	if (user){
		if (this.populated("_usersAttending")){
			for (var i = 0; i < this._usersAttending.length; i++){
				var curr = this._usersAttending[i];

				if (curr._id.equals(user._id)){
					return true;
				}
			}
		}

		else{
			return this._usersAttending.indexOf(user._id) != -1;
		}
	}

	return false;
};

soireeSchema.methods.chargedForReservation = function(reservation){
	var soiree = this;

	console.log("in chargedForReservation()");

	reservation.deepPopulate("_business _user", function(err, _reservation){
		//reservation has now been charged. Remove from unchargedReservations
		ArrayHelper.removeObjectPopulated(soiree, "_unchargedReservations", _reservation);

		//remove user from list of users that we havent charged yet
		ArrayHelper.removeObjectPopulated(soiree, "_usersUncharged", _reservation._user);

		//add reservation to soirees chargedReservations
		ArrayHelper.pushOnlyOncePopulated(soiree, "_chargedReservations", _reservation);

		//add user to soirees usersAttending
		ArrayHelper.pushOnlyOncePopulated(soiree, "_usersAttending", _reservation._user);

		//add reservation to business's unconfirmedReservations
		ArrayHelper.pushOnlyOncePopulated(_reservation._business, "_unconfirmedReservations", _reservation);

		soiree.save(Globals.saveErrorCallback);
		_reservation._business.save(Globals.saveErrorCallback);
	});


};

soireeSchema.methods.join = function(user, successCallback, errorCallback){
	var soiree = this;

	if (soiree.college !== user.college) return errorCallback(ErrorCodes.InvalidInput);

	if (!this.full){
		//var stripeToken = req.body.stripeToken;
		//if (!stripeToken){
		//	return ResHelper.sendError(res, errorCodes.MissingStripeToken);
		//}
		if (!user.stripeCustomerId && !(user.testUser && Globals.development)){
			return errorCallback(ErrorCodes.MissingStripeCustomerId);
		}
		if (this._usersAttending.indexOf(user._id) != -1 || this._usersUncharged.indexOf(user._id) != -1){
			//user has already joined soiree
			return errorCallback(ErrorCodes.UserAlreadyJoinedSoiree);
		}

		else{
			//user can join
			//see if you should charge him now
			var SoireeReservation = mongoose.model("SoireeReservation");
			if (this.willReachNumUsersMin || this.reachedNumUsersMin){
				if (this._unchargedReservations.length > 0 || this._usersUncharged.length > 0){
					this.chargeUnchargedUsers();
				}
				SoireeReservation.createChargedSoireeReservation(user, soiree, successCallback, errorCallback);
			}
			else{
				SoireeReservation.createUnchargedSoireeReservation(user, soiree, successCallback, errorCallback);
			}
		}

	}
	else{
		return errorCallback(ErrorCodes.SoireeFull);
	}
};

soireeSchema.methods.chargeUnchargedUsers = function(){

	console.log("In chargeUnchargedUses() : Charging " + this._unchargedReservations.length + " users...");
	console.log(this._unchargedReservations);

	var soiree = this;
	this.deepPopulate("_unchargedReservations._user", function(err, _soiree){
		if (err){
			return console.log("ERROR POPULATING: " + err);
		}
		var shadowCopy = _soiree._unchargedReservations;
		for (var i = 0; i < shadowCopy.length; i++){
			var reservation = shadowCopy[i];
			reservation.chargeUser(function(user){
				console.log("Successfully charged " + user.fullName);
				soiree.chargedForReservation(reservation);
				//console.log(_soiree);
			}, function(err){
				console.log("Error charging user " + reservation._user.fullName + " : " + err);
			});
		}
		console.log(_soiree._unchargedReservations);
	});

};

soireeSchema.methods.jsonObject = function (user) {
	var timeIntervalSince1970InSeconds = this.date.getTime() / 1000;

	var usersColleges = [];
	//console.log("_usersAttending: " + this._usersAttending);

	for (var i = 0; i < this._usersAttending.length; i++){
		var college = this._usersAttending[i].college;
		if (college){
			usersColleges.push(college);
		}
	}

	var obj = {
		"soireeType": this.soireeType,
		"numUsersAttending": this.numUsersAttending,
		"numUsersMax": this.numUsersMax,
		"date": timeIntervalSince1970InSeconds,
		"soireeId": this.soireeId,
		"started" : this.started,
		"ended" : this.ended,
		"inProgress" : this.inProgress,
		"businessName": this._business.businessName,
		"cityArea" : this._business.cityArea,
		"coordinates" : this.location.coordinates,
		"initialCharge": this.initialCharge,
		"photoIndexIdentifier" : this.photoIndexIdentifier,
		"reachedNumUsersMin" : this.reachedNumUsersMin,
		"willReachNumUsersMin" : this.willReachNumUsersMin,
		"soireeDescription" : this.soireeDescription,
		"numUsersMin" : this.numUsersMin,
		"open" : this.open,
		"college" : this.college
	};

	if (user){
		obj.userAlreadyJoined = this.hasUserAlreadyJoined(user);
	}

	return obj;
};

soireeSchema.methods.createDescription = function(){
	var soireeType = this.soireeType.toLowerCase();
	var description = "";

	var numOtherUsers = (this.numUsersMax-1) + "";

	if (soireeType === "lunch"){
		description =
				"Come grab lunch with " + numOtherUsers +
				" other amazing people! Eat delicious food while " +
				"playing games and chatting with soon-to-be new friends. Not sure how to break the ice? " +
				"We'll do all the work, don't worry.";
	}
	else if (soireeType === "dinner"){
		description =
			"Come grab dinner with " + numOtherUsers +
			" other amazing people! Eat delicious food while " +
			"playing games and chatting with soon-to-be new friends. Not sure how to break the ice? " +
			"We'll do all the work, don't worry.";

	}
	else if (soireeType === "drinks"){
		description =
			"Come grab drinks with " + numOtherUsers +
			" other amazing people! Drink quality alcoholic drinks while " +
			"playing games and chatting with soon-to-be new friends. Not sure how to break the ice? " +
			"We'll do all the work, don't worry.";
	}
	else if (soireeType === "blind date"){
		description =
			"Want to meet a new love interest, but not sure how? Worried about all the creeps out there," +
			"or about the date being boring? Try a " + this.SOIREE + " blind date! " +
			"We'll make sure the person on the other end is not a creep, and the date is exciting. Not sure how to break the ice? " +
			"We'll do all the work, don't worry.";
	}
	return description;
};

function determineDateForSoireeType(soireeType, college, callback){
	console.log('date');

	var dateCallback = function(counts){
		console.log(counts);

		if (!counts || counts.length === 0) return callback(null);

		var minDate = new Date(Date.now() + MIN_TIME_BEFORE_CAN_CREATE_SOIREE);

		var availableTimes = Globals.availableTimesPerSoireeType[soireeType];

		var minTime;
		var min;

		if (minDate <= DateHelper.dateFromTime(availableTimes[0])) {
			minTime = availableTimes[0];
			min = counts[minTime];
		}

		for (var i = 1; i < availableTimes.length; i++){
			var curr = availableTimes[i];

			console.log('comparing ' + counts[curr] + ' with ' + min);

			if ((!min || counts[curr] < min) && minDate <= DateHelper.dateFromTime(curr)){
				min = counts[curr];
				minTime = curr;
			}
		}

		if (!minTime) return callback(null);

		var ans = DateHelper.dateFromTime(minTime);
		console.log("Date determined: " + ans.toString());
		return callback(ans);
	};


	if (!timeCountsPerCollege[college][soireeType]){
		countsForSoireeTypeAtTimes(soireeType, college, dateCallback);
	}
	else{
		var counts = timeCountsPerCollege[college][soireeType];
		dateCallback(counts);
	}




};

function alertUsersThatSoireeStarted(soiree){
	for (var i = 0; i < soiree._usersAttending.length; i++){
		var user = soiree._usersAttending[i];
		console.log("Sending push notification to " + user.firstName);

		var message = "Your " + soiree.soireeType + " " + soiree.SOIREE_LOWERCASE + " is about to start! Open up " + soiree.SOIREE + " to get started.";
		PushNotificationHelper.sendPushNotification(user, message);
	}
};


/* Virtuals */

soireeSchema.virtual('full').get(function () {
	return this.numUsersAttending >= this.numUsersMax;
});

soireeSchema.virtual('SOIREE').get(function () {
	return "Soirée";
});

soireeSchema.virtual('reachedNumUsersMin').get(function () {
	//var min = this.numUsersMax/2 + (this.numUsersMax % 2);
	return this.totalNumUsers >= this.numUsersMin;
});

//soireeSchema.virtual('numUsersMin').get(function () {
//	if (this.numUsersMax <= 2){
//		return this.numUsersMax;
//	}
//	return parseInt(this.numUsersMax/2) + this.numUsersMax%2; //rounds down, then round up by adding 1 if necessary
//});

soireeSchema.virtual('willReachNumUsersMin').get(function () {
	return this.totalNumUsers === this.numUsersMin-1;
});

soireeSchema.virtual('numUsersAttending').get(function () {
	return this._usersAttending.length;
});

soireeSchema.virtual('totalNumUsers').get(function () {
	return this._usersAttending.length + this._usersUncharged.length;
});

soireeSchema.virtual('inProgress').get(function () {
	return this.started && !this.ended;
});


soireeSchema.pre("save", function(next){

	if (!this.scheduledStartTimeIdentifier) {
		this.scheduledStartTimeIdentifier = this.constructor.createScheduledTimeIdentifier(this.date);
	}
	if (!this.scheduledEndTimeIdentifier){
		this.scheduledEndTimeIdentifier = this.constructor.createScheduledTimeIdentifierFuture(this.length);
	}
	if (!this.numUsersMax){
		this.numUsersMax = Globals.numUsersMaxPerSoireeType[this.soireeType];
	}
	if (!this.initialCharge){
		this.initialCharge = Globals.initialChargePerSoireeType[this.soireeType];
	}
	if (!this.soireeDescription){
		this.soireeDescription = this.createDescription();
	}
	//if (!this.photoIndexIdentifier){
	//	this.photoIndexIdentifier = generatePhotoIndexIdentifier();
	//}
	//console.log("this.duration: " + this.duration);
5
	next();
});

soireeSchema.post("init", function(soiree){

});

//var autoPopulate = function(next){
//	this.populate("_business");
//	this.populate("_user");
//	next();
//};
//
//soireeSchema.pre("findOne", autoPopulate);
//soireeSchema.pre("find", autoPopulate);

//soireeSchema.pre('save', function(next){
//	//if (!this.timeAtString) {
//	this.timeAtString = soireeSchema.statics.timeAtStringFromDate(this.date);
//	//}
//	next();
//});

var deepPopulate = require('mongoose-deep-populate')(mongoose);
var options = {};
soireeSchema.plugin(deepPopulate, options);

var SoireeModel = mongoose.model('Soiree', soireeSchema);
module.exports = SoireeModel;

SoireeModel.fillTimeCountsPerCollege();
