// app/db/SoireeReservation.js
var mongoose = require('./mongoose_connect.js');
var Schema = mongoose.Schema;
var ObjectId = Schema.Types.ObjectId;


/* Other Models */
var Business = require('./Business.js');
var User = require('./User.js');
var Soiree = require('./Soiree.js');

/* Packages */
var shortid = require('shortid');
var _ = require("underscore");
var Q = require('Q');

/* Helper */
var helpersFolderLocation = "../../helpers/";
var DateHelper = require('app/helpers/DateHelper.js');
var ResHelper = require('app/helpers/ResHelper.js');
var CreditCardHelper = require('app/helpers/CreditCardHelper.js');
var LocationHelper = require('app/helpers/LocationHelper.js');
var Globals = require('app/helpers/Globals.js');
var ArrayHelper = require('app/helpers/ArrayHelper.js');
var PushNotificationHelper = require('app/helpers/PushNotificationHelper.js');
var MongooseHelper = require('app/helpers/MongooseHelper.js');
var IDGeneratorHelper = require('app/helpers/IDGeneratorHelper.js');

var ErrorCodes = require('app/helpers/ErrorCodes.js');

var customSchema = new Schema({
        _user : {type: ObjectId, ref:"User", required: true},
        _soiree : {type: ObjectId, ref: "Soiree", required: true},
        _business : {type: ObjectId, ref: "Business", required: true},
        reservationId: {type: String, index: true, unique: true, required: true},
        soireeId: {type: String},
        charged : {type: Boolean, default: false},
        charge : {type: Schema.Types.Mixed},
        confirmationCode : {type: String, uppercase: true, trim: true},
        confirmed : {type: Boolean, default: false},
        confirmationCodeSet : {type: Boolean, default: false},
        date : {type: Date}

    },
    {timestamps: {createdAt: 'dateCreated', updatedAt: 'dateUpdated'}}
);


//customSchema.index({ confirmationCode: 1, _business: 1, confirmed: 1}, { unique: true }, function(err, indexName){
//   console.log("index callback with err: " + err + " index: " + indexName);
//});

function generateConfirmationCode(){
    return IDGeneratorHelper.generateId(3);
};

//customSchema.statics.chargedForReservations = function (reservations) {
//    for (var i = 0; i < reservations.length; i++) {
//        reservations[i].successfullyCharged();
//    }
//};

//called when a reservation wasnt charged before but now is
customSchema.methods.chargedUnchargedReservation = function () {
    this.deepPopulate("_soiree _business _user", function(err, reservation){
        if (err){
            console.log(err);
            return;
        }

        reservation._user.chargedUnchargedReservation(reservation);
        reservation._business.chargedUnchargedReservation(reservation);
        reservation._soiree.chargedUnchargedReservation(reservation);
    });

};

customSchema.statics.createUnchargedSoireeReservation = function(user, soiree, successCallback, errorCallback){
    var business = soiree._business;

    var reservation = new this({
        _user : user._id,
        _soiree : soiree._id,
        soireeId : soiree.soireeId,
        _business : business._id,
        charged : false,
        date : soiree.date
    });

    reservation.save(function(err, _reservation) {
        if (err) {
            console.log("Error saving reservation: " + err);
            errorCallback(ErrorCodes.MongoError);
        }
        else{

            //var promises = [
                user.addNewReservation(_reservation),
                soiree.addNewReservation(_reservation),
                business.addNewReservation(_reservation),
            //];

            //Q.all(promises).then(function(){
                successCallback(soiree);
            //}).fail(function(err){
            //    console.log(err);
            //    errorCallback(ErrorCodes.ErrorSaving);
            //});

        }
    });
};



customSchema.statics.createChargedSoireeReservation = function(user, soiree, successCallback, errorCallback) {
    if (soiree._chargedReservations.indexOf(user._id) != -1){
        return errorCallback(ErrorCodes.AlreadyExists);
    }
    else if (soiree._unchargedReservations.indexOf(user._id) != -1){
        return errorCallback(ErrorCodes.AlreadyExists);
    }
    else{
        var SoireeReservation = this;

        console.log("In createChargedSoireeReservation() : charging " + user.fullName);
        soiree.deepPopulate("_business", function(err){
            if (err){
                console.log(err);
                return errorCallback(ErrorCodes.MongoError);
            }
            var business = soiree._business;

            CreditCardHelper.chargeForSoiree(soiree, user, function(charge) {

                if (!charge){ return errorCallback(ErrorCodes.PaymentError); }

                var reservation = new SoireeReservation({
                    _user : user._id,
                    _soiree : soiree._id,
                    soireeId : soiree.soireeId,
                    _business : business._id,
                    charge : charge,
                    charged : true,
                    date : soiree.date
                });


                reservation.save(function(err, _reservation){
                    if (err){
                        console.log("Error saving reservation: " + err);
                        errorCallback(ErrorCodes.ErrorSaving);
                    }
                    else{
                        //var promises = [
                             user.addNewReservation(_reservation),
                             soiree.addNewReservation(_reservation),
                             business.addNewReservation(_reservation),
                        //];

                        //Q.all(promises).then(function(){
                            successCallback(soiree);
                        //}).fail(function(err){
                        //    console.log(err);
                        //    errorCallback(ErrorCodes.ErrorSaving);
                        //});


                    }
                });


            }, function(err){
                errorCallback(err);
            });
        });

    }
};


customSchema.methods.chargeUser = function(successCallback, errorCallback){
    if (!this.charged){
        console.log("about to charge " + this._user.fullName);

        this.deepPopulate("_soiree _user", function(err, _reservation){
            CreditCardHelper.chargeForSoiree(_reservation._soiree, _reservation._user, function(charge) {
                if (!charge){ return errorCallback(ErrorCodes.PaymentError); }

                _reservation.charged = true;
                _reservation.charge = charge;

                _reservation.save(function(err){
                    if (err){
                        console.log(err);
                        //TODO: Refund Charge
                        errorCallback(err);
                    }
                    else{
                        _reservation.chargedUnchargedReservation();
                        successCallback(_reservation._user);
                    }
                });
            }, function(err){
                errorCallback(err);
            });
        });
    }

};


customSchema.statics.checkIfConfirmationCodeIsUnique = function(reservation, successCallback, errorCallback){
    SoireeReservation.find({confirmationCode : reservation.confirmationCode, _business : reservation._business, confirmed: false}).exec(function(err, reservations){
        if (err){
            errorCallback(err);
        }
        else{
            if (reservations.length === 0){
                successCallback(true);
            }
            else{
                successCallback(false);
            }
        }
    });
};

customSchema.methods.confirm = function(code, successCallback, errorCallback){
    code = code.toUpperCase();

    var reservation = this;

    if (code === this.confirmationCode && !this.confirmed){
        this.confirmed = true;
        this.confirmationCode = "ALREADY_CONFIRMED";

        //console.log(this);

        this.save(function(err){
            if (err){
                console.log(err);
                errorCallback(ErrorCodes.ErrorSaving);
            }
            else{
                console.log("confirmed. running business tasks...");
                reservation.populate("_business", function(err){
                    if (err){
                        console.log("error populating");

                        console.log(err);
                        return errorCallback(ErrorCodes.ErrorSaving);
                    }
                    reservation._business.confirmSoireeReservation(reservation, successCallback, errorCallback);
                });
                //console.log("Successfully confirmed reservation");
                //successCallback();
            }
        });
    }
    else errorCallback();
};

customSchema.statics.addReservationsForSoirees = function(soirees, user, successCallback){
    if (soirees.length === 0){
        return successCallback({});
    }

    var ans = {};
    var numReturned = 0;

    for (var i = 0; i < soirees.length; i++){
        var soireeId = soirees[i].soireeId;
        this.findOne({soireeId: soireeId, _user: user._id}).exec(function (err, reservation) {
            numReturned++;
            if (err || !reservation) {
                console.log(err);
            }
            else {
                ans[soireeId] = reservation.jsonObject();
            }
            if (numReturned === soirees.length){
                successCallback(ans);
            }
        });
    }
};

customSchema.methods.jsonObject = function(){
    console.log("reservation: " + this);

  var obj = {
      "reservationId" : this.reservationId,
      "soireeId" : this.soireeId,
      "confirmationCode" : this.confirmationCode,
      "confirmed" : this.confirmed,
      "charged" : this.charged
  };

    if (this._business && this._business.businessName){
        obj["businessName"] = this._business.businessName;
    }

    return obj;
};
customSchema.methods.refund = function(){
  //TODO: implement soiree reservation refund
};

customSchema.virtual('').get(function () {
});

customSchema.pre("validate", function (next) {
    if (!this.reservationId){
        IDGeneratorHelper.generateUniqueId("SoireeReservation", "reservationId", this, next, {length : 15, addLowerCase : true});
    }
    else next();
});

customSchema.pre("save", function (next) {

    if (!this.confirmationCodeSet && !this.confirmed  && this.charged && this.charge){
        var retries = 0;

        var reservation = this;
        reservation.confirmationCode = generateConfirmationCode();

        var check = function(){
            SoireeReservation.checkIfConfirmationCodeIsUnique(reservation, function(unique){
                if (unique){
                    console.log(reservation.confirmationCode + " worked. ");
                    this.confirmationCodeSet = true;
                    next();
                }
                else{
                    console.log(reservation.confirmationCode + " didnt work. ");
                    if (retries < Globals.RETRIES_ON_DUPLICATE_CONFIRMATION_CODE){
                        reservation.confirmationCode = generateConfirmationCode();
                        retries++;
                        check();
                    }
                    else next(new Error("Unable to create unique confirmation code"));
                }
            }, function(err){
                console.log("Error checking if there are any existing reservations with same confirmation code : " + err);
                next(new Error("Error checking for existing reservations in pre(save)"));
            });
        };

        check();
    }
    else next();

});


customSchema.post("init", function (soiree) {

});

var autoPopulate = function(next){
    this.populate("_business");
    this.populate("_soiree");
    this.populate("_user");
    next();
};

customSchema.pre("findOne", autoPopulate);
customSchema.pre("find", autoPopulate);

var deepPopulate = require('mongoose-deep-populate')(mongoose);
var options = {};
customSchema.plugin(deepPopulate, options);


var SoireeReservation = mongoose.model('SoireeReservation', customSchema);
module.exports = SoireeReservation;
