// app/db/UserVerification.js
/**
 * Created by shadygabal on 11/26/15.
 */
/* Setup */
var mongoose = require('./mongoose_connect.js');
var Schema = mongoose.Schema;
var ObjectId = Schema.Types.ObjectId;


/* Other Models */
var Business = require('./Business.js');
var Soiree = require('./Soiree.js');
var User = require('./User.js');
var Admin = require('./Admin.js');
var Image = require('./Image.js');
var Globals = require('app/helpers/Globals');

/* Packages */
var shortid = require('shortid');


var userVerificationSchema = new Schema({
        _idImage : {type: ObjectId, ref: "Image"},
        _selfImage : {type: ObjectId, ref:"Image"},
        idImageUrl : {type: String},
        selfImageUrl : {type: String},
        userVerificationId: {type: String, index: true, default: shortid.generate},
        _user : {type: ObjectId, ref: "User"},
        notes : {type: String},
        //college : {type: String, enum: Globals.colleges},
        accepted : {type : Boolean, default : false},
        rejected : {type: Boolean, default: false},
        handled : {type: Boolean, default: false, index:true},
        dateVerified : {type: Date},
        _approvedBy: {type: ObjectId, ref: "Admin"}

},
    { timestamps: { createdAt: 'dateCreated', updatedAt: 'dateUpdated' } }
);


userVerificationSchema.statics.findUnverifiedVerifications = function(admin, idsToIgnore, successCallback, errorCallback){
    if (admin){
        var UserVerification = this;

        var constraints = {handled: false};

        if (idsToIgnore && idsToIgnore.length > 0){
            constraints["_id"] = {'$nin' : idsToIgnore};
        }

        this.find(constraints).deepPopulate("_user").limit(10).exec(function(err, verifications){
            if (err){
                console.log("Error finding unverifiedVerifications - findUnverifiedVerifications(): " + err);
                errorCallback(err);
            }
            else{
                //console.log("Found verifications: " + verifications);
                successCallback(verifications);
            }

        });
    }
    else{
        errorCallback();
    }

};

userVerificationSchema.methods.accept = function(admin, successCallback, errorCallback){
    var verification = this;
    if (verification.handled){
        return errorCallback(ErrorCodes.InvalidInput);
    }

    verification.deepPopulate("_user", function(err){
       if (err){
           console.log("Error deep populating verification: " + verification);
           errorCallback(ErrorCodes.MongoError);
       }
        else{
           verification._user.verified = true;
           verification._user.pendingVerification = false;
           verification.accepted = true;
           verification.handled = true;
           verification.save(function(err){
              if (err){
                  console.log("Error saving user after verifying: " + err);
                  errorCallback(ErrorCodes.MongoError);
              }
               else{
                  verification._user.notifyOfVerification();
                  successCallback();
              }
           });

       }
    });
};

userVerificationSchema.methods.reject = function(admin, reason, successCallback, errorCallback){
    var verification = this;
    if (verification.handled){
        return errorCallback(ErrorCodes.InvalidInput);
    }

    verification.deepPopulate("_user", function(err){
        if (err){
            console.log("Error deep populating verification: " + verification);
            errorCallback(ErrorCodes.MongoError);
        }
        else{
            verification._user.verified = false;
            verification._user.pendingVerification = false;
            verification.rejected = true;
            verification.handled = true;
            verification.save(function(err){
                if (err){
                    console.log("Error saving user after verifying: " + err);
                    errorCallback(ErrorCodes.MongoError);
                }
                else{
                    verification._user.notifyOfVerification(reason);
                    successCallback();
                }
            });

        }
    });
};

userVerificationSchema.post('remove', function(doc) {
    Image.remove({_userVerification : doc._id}).exec(function(err){
       if (err){
           console.log("Error removing ghost images in Image.post('remove'): " + err + "\n POSSIBLE DATA LEAK");
       }
    });

});

var deepPopulate = require('mongoose-deep-populate')(mongoose);
var options = {};
userVerificationSchema.plugin(deepPopulate, options);

module.exports = mongoose.model('UserVerification', userVerificationSchema);
